mesonpy/_wheelfile.py

```diff
@@ -0,0 +1,99 @@
+# SPDX-License-Identifier: MIT
+
+import base64
+import csv
+import hashlib
+import io
+import os
+import re
+import stat
+import time
+import zipfile
+
+from types import TracebackType
+from typing import List, Optional, Tuple, Type, Union
+
+
+Path = Union[str, os.PathLike]
+
+
+MIN_TIMESTAMP = 315532800  # 1980-01-01 00:00:00 UTC
+WHEEL_FILENAME_REGEX = re.compile(r'^(?P<name>[^-]+)-(?P<version>[^-]+)(:?-(?P<build>[^-]+))?-(?P<tag>[^-]+-[^-]+-[^-]+).whl$')
+
+
+def _b64encode(data: bytes) -> bytes:
+    return base64.urlsafe_b64encode(data).rstrip(b'=')
+
+
+class WheelFile:
+    """Implement the wheel package binary distribution format.
+
+    https://packaging.python.org/en/latest/specifications/binary-distribution-format/
+    """
+    def __new__(cls, filename: Path, mode: str = 'r', compression: int = zipfile.ZIP_DEFLATED) -> 'WheelFile':
+        if mode == 'w':
+            return super().__new__(WheelFileWriter)
+        raise NotImplementedError
+
+    @staticmethod
+    def timestamp(mtime: Optional[float] = None) -> Tuple[int, int, int, int, int, int]:
+        timestamp = int(os.environ.get('SOURCE_DATE_EPOCH', mtime or time.time()))
+        timestamp = max(timestamp, MIN_TIMESTAMP)
+        return time.gmtime(timestamp)[0:6]
+
+    @staticmethod
+    def hash(data: bytes) -> str:
+        return 'sha256=' + _b64encode(hashlib.sha256(data).digest()).decode('ascii')
+
+    def writestr(self, zinfo_or_arcname: Union[str, zipfile.ZipInfo], data: bytes) -> None:
```

Comment:
`mypy` is not happy with this type annotation
```
mesonpy/__init__.py:505: error: Argument 2 to "writestr" of "WheelFile" has incompatible type "_lru_cache_wrapper[bytes]"; expected "bytes"  [arg-type]
```
but my Python typing experience is not enough to decide what's the correct fix. Ideas?

Reply:
I solved this in the second commit in this PR. I think that this is a bug in `mypy` but the code in `meson-python` was a bit unusual without reason, so I changed it.